components/dashboard/src/projects/new-project/NewProjectRepoSelection.tsx

```diff
@@ -19,38 +19,47 @@ import { useToast } from "../../components/toasts/Toasts";
 import { useProviderRepositoriesForUser } from "../../data/git-providers/provider-repositories-query";
 import { openReconfigureWindow } from "./reconfigure-github";
 import { NewProjectCreateFromURL } from "./NewProjectCreateFromURL";
+import { useStateWithDebounce } from "../../hooks/use-state-with-debounce";
+import { useFeatureFlag } from "../../data/featureflag-query";
 
 type Props = {
-    selectedProviderHost?: string;
+    selectedProvider?: AuthProviderInfo;
     onProjectCreated: (project: Project) => void;
     onChangeGitProvider: () => void;
 };
-export const NewProjectRepoSelection: FC<Props> = ({ selectedProviderHost, onProjectCreated, onChangeGitProvider }) => {
+export const NewProjectRepoSelection: FC<Props> = ({ selectedProvider, onProjectCreated, onChangeGitProvider }) => {
     const { toast } = useToast();
+    const newProjectIncrementalRepoSearchBBS = useFeatureFlag("newProjectIncrementalRepoSearchBBS");
     const { data: isGitHubAppEnabled } = useIsGithubAppEnabled();
-    const areGitHubWebhooksUnauthorized = useAreGithubWebhooksUnauthorized(selectedProviderHost || "");
+    const areGitHubWebhooksUnauthorized = useAreGithubWebhooksUnauthorized(selectedProvider?.host ?? "");
     const createProject = useCreateProject();
 
     // Component state managed by this component
     const [selectedAccount, setSelectedAccount] = useState<string>();
-    const [repoSearchFilter, setRepoSearchFilter] = useState("");
+    const [repoSearchFilter, setRepoSearchFilter, debouncedRepoSearchFilter] = useStateWithDebounce("");
     const [installationId, setInstallationId] = useState<string>();
 
     // Main query for listing repos given the current state
     const { data: reposInAccounts, isLoading } = useProviderRepositoriesForUser({
-        provider: selectedProviderHost || "",
+        providerHost: selectedProvider?.host ?? "",
         installationId,
+        search: debouncedRepoSearchFilter,
     });
 
     // Wrap setting selected account so we can clear the repo search filter at the same time
-    const setSelectedAccountAndClearSearch = useCallback((account?: string) => {
-        setSelectedAccount(account);
-        setRepoSearchFilter("");
-    }, []);
+    const setSelectedAccountAndClearSearch = useCallback(
+        (account?: string) => {
+            setSelectedAccount(account);
+            setRepoSearchFilter("");
+        },
+        [setRepoSearchFilter],
+    );
 
     // Memoized & derived values
     const noReposAvailable = !!(reposInAccounts?.length === 0 || areGitHubWebhooksUnauthorized);
-    const isGitHub = selectedProviderHost === "github.com";
+    const isGitHub = selectedProvider?.host === "github.com";
+    const isBitbucketServer = selectedProvider?.authProviderType === "BitbucketServer";
+    const enableIncrementalSearch = isBitbucketServer && newProjectIncrementalRepoSearchBBS;
```

Comment:
(nit) `enableBBSIncrementalSearch`

Reply:
Good point - I was assuming we'd want to do this same thing for other git providers at some point since loading all repos and filtering on the client will be a problem for any of them. Renaming for now will make it a bit more clear though.